Compiler error when using a string generated with the `fmt""` macro twice in a template.

### What happened?

If I pass a string generated using the `fmt""` macro from `strformat` to a template that accepts a string, and then use that string twice in the template, I get a C compiler error. Here's a minimal example:

```nim
import strformat

template myTemplate(s: string) =
  echo s
  echo s

proc main() =
  myTemplate fmt"hello"

main()
```

When trying to compile this, I get an error message from the C compiler step shown below. If I don't use the `fmt""` macro, this compiles as expected, and if I remove the the `main()` function and just call `myTemplate` globally it also compiles. Furthermore, the issue remains if I change the template argument type to `typed`, but it disappears if I change it to `untyped`.

### Nim Version

Tested with 1.6.6 and 1.2.0 with the same result.

### Current Standard Output Logs

```shell
Hint: used config file '/home/XXX/.choosenim/toolchains/nim-1.6.6/config/nim.cfg' [Conf]
Hint: used config file '/home/XXX/.choosenim/toolchains/nim-1.6.6/config/config.nims' [Conf]
..........................................................................
CC: template_test.nim
/home/XXX/.cache/nim/template_test_d/@mtemplate_test.nim.c: In function ‘main__template95test_4’:
/home/XXX/.cache/nim/template_test_d/@mtemplate_test.nim.c:118:24: error: redeclaration of ‘fmtRes’ with no linkage
  118 |         NimStringDesc* fmtRes;
      |                        ^~~~~~
/home/XXX/.cache/nim/template_test_d/@mtemplate_test.nim.c:116:24: note: previous declaration of ‘fmtRes’ with type ‘NimStringDesc *’
  116 |         NimStringDesc* fmtRes;
      |                        ^~~~~~
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3   -I/home/XXX/.choosenim/toolchains/nim-1.6.6/lib -I/home/XXX/projects/nim -o /home/XXX/.cache/nim/template_test_d/@mtemplate_test.nim.c.o /home/XXX/.cache/nim/template_test_d/@mtemplate_test.nim.c' failed with exit code: 1
```


### Expected Standard Output Logs

```shell
hello
hello
```


### Possible Solution

I have no solution, but the issue can be worked around by either using specifying the template argument as `untyped` or by using a temporary variable in the template:

```nim
template myTemplate(s: string) =
  let s2 = s
  echo s2
  echo s2
```

### Additional Information

As far as I can tell, this is a different issue than the limitation with `strformat` discussed here: https://nim-lang.org/docs/strformat.html#limitations

--- a/lib/pure/strformat.nim
+++ b/lib/pure/strformat.nim
@@ -577,7 +577,8 @@ template formatValue(result: var string; value: char; specifier: string) =
 template formatValue(result: var string; value: cstring; specifier: string) =
   result.add value
 
-proc strformatImpl(f: string; openChar, closeChar: char): NimNode =
+proc strformatImpl(f: string; openChar, closeChar: char,
+                   lineInfoNode: NimNode = nil): NimNode =
   template missingCloseChar =
     error("invalid format string: missing closing character '" & closeChar & "'")
 
@@ -585,7 +586,7 @@ proc strformatImpl(f: string; openChar, closeChar: char): NimNode =
     error "openChar and closeChar must not be ':'"
   var i = 0
   let res = genSym(nskVar, "fmtRes")
-  result = newNimNode(nnkStmtListExpr)
+  result = newNimNode(nnkStmtListExpr, lineInfoNode)
   # XXX: https://github.com/nim-lang/Nim/issues/8405
   # When compiling with -d:useNimRtl, certain procs such as `count` from the strutils
   # module are not accessible at compile-time:
@@ -644,6 +645,7 @@ proc strformatImpl(f: string; openChar, closeChar: char): NimNode =
           x = parseExpr(subexpr)
         except ValueError as e:
           error("could not parse `$#` in `$#`.\n$#" % [subexpr, f, e.msg])
+        x.copyLineInfo(lineInfoNode)
         let formatSym = bindSym("formatValue", brOpen)
         var options = ""
         if f[i] == ':':
@@ -672,7 +674,14 @@ proc strformatImpl(f: string; openChar, closeChar: char): NimNode =
   when defined(debugFmtDsl):
     echo repr result
 
-macro fmt*(pattern: static string; openChar: static char, closeChar: static char): string =
+macro fmt(pattern: static string; openChar: static char, closeChar: static char, lineInfoNode: untyped): string =
+  ## version of `fmt` with dummy untyped param for line info
+  strformatImpl(pattern, openChar, closeChar, lineInfoNode)
+
+when not defined(nimHasCallsitePragma):
+  {.pragma: callsite.}
+
+template fmt*(pattern: static string; openChar: static char, closeChar: static char): string {.callsite.} =
   ## Interpolates `pattern` using symbols in scope.
   runnableExamples:
     let x = 7
@@ -689,13 +698,13 @@ macro fmt*(pattern: static string; openChar: static char, closeChar: static char
     assert "<x>".fmt('<', '>') == "7"
     assert "<<<x>>>".fmt('<', '>') == "<7>"
     assert "`x`".fmt('`', '`') == "7"
-  strformatImpl(pattern, openChar, closeChar)
+  fmt(pattern, openChar, closeChar, dummyForLineInfo)
 
-template fmt*(pattern: static string): untyped =
+template fmt*(pattern: static string): untyped {.callsite.} =
   ## Alias for `fmt(pattern, '{', '}')`.
-  fmt(pattern, '{', '}')
+  fmt(pattern, '{', '}', dummyForLineInfo)
 
-macro `&`*(pattern: string{lit}): string =
+template `&`*(pattern: string{lit}): string {.callsite.} =
   ## `&pattern` is the same as `pattern.fmt`.
   ## For a specification of the `&` macro, see the module level documentation.
   # pending bug #18275, bug #18278, use `pattern: static string`
@@ -707,4 +716,4 @@ macro `&`*(pattern: string{lit}): string =
     assert &"{x}\n" == "7\n" # regular string literal
     assert &"{x}\n" == "{x}\n".fmt # `fmt` can be used instead
     assert &"{x}\n" != fmt"{x}\n" # see `fmt` docs, this would use a raw string literal
-  strformatImpl(pattern.strVal, '{', '}')
+  fmt(pattern, '{', '}', dummyForLineInfo)

--- a/lib/pure/strformat.nim
+++ b/lib/pure/strformat.nim
@@ -671,6 +671,11 @@ proc strformatImpl(f: string; openChar, closeChar: char,
   if strlit.len > 0:
     result.add newCall(bindSym"add", res, newLit(strlit))
   result.add res
+  # workaround for #20381
+  var blockExpr = newNimNode(nnkBlockExpr, lineInfoNode)
+  blockExpr.add(newEmptyNode())
+  blockExpr.add(result)
+  result = blockExpr
   when defined(debugFmtDsl):
     echo repr result
 

--- a/tests/stdlib/tstrformat.nim
+++ b/tests/stdlib/tstrformat.nim
@@ -562,6 +562,16 @@ proc main() =
     doAssert &"""{(if true: "'" & "'" & ')' else: "")}""" == "'')"
     doAssert &"{(if true: \"\'\" & \"'\" & ')' else: \"\")}" == "'')"
     doAssert fmt"""{(if true: "'" & ')' else: "")}""" == "')"
+  
+  block: # issue #20381
+    var ss: seq[string]
+    template myTemplate(s: string) =
+      ss.add s
+      ss.add s
+    proc foo() =
+      myTemplate fmt"hello"
+    foo()
+    doAssert ss == @["hello", "hello"]
 
 static: main()
 main()

--- a/tests/stdlib/tstrformatlineinfo.nim
+++ b/tests/stdlib/tstrformatlineinfo.nim
@@ -1,0 +1,8 @@
+# issue #21759
+
+{.hint[ConvFromXToItselfNotNeeded]: on.}
+
+import std/strformat
+
+echo fmt"{string ""abc""}" #[tt.Hint
+        ^ conversion from string to itself is pointless]#

